FWCore/Framework/src/ProductResolvers.cc

```diff
@@ -665,21 +648,26 @@ namespace edm {
   void SwitchBaseProductResolver::resetProductData_(bool deleteEarly) {
     productData_.resetProductData();
     realProduct_.resetProductData_(deleteEarly);
-    if (not deleteEarly) {
-      status_ = defaultStatus_;
-    }
+    prefetchRequested_ = false;
+    waitingTasks_.reset();
```

Comment:
If `deleteEarly == true` then neither of these items should be reset.

Reply:
Other ProductResolvers seem to reset these variables unconditionally
https://github.com/cms-sw/cmssw/blob/4010528c5c1660debe651b9f69d474a7f8b5004b/FWCore/Framework/src/ProductResolvers.cc#L295-L297
https://github.com/cms-sw/cmssw/blob/4010528c5c1660debe651b9f69d474a7f8b5004b/FWCore/Framework/src/ProductResolvers.cc#L359-L362
https://github.com/cms-sw/cmssw/blob/4010528c5c1660debe651b9f69d474a7f8b5004b/FWCore/Framework/src/ProductResolvers.cc#L450-L452
https://github.com/cms-sw/cmssw/blob/4010528c5c1660debe651b9f69d474a7f8b5004b/FWCore/Framework/src/ProductResolvers.cc#L1026-L1032

Also, shouldn't the early deletion be done only after all consumers of the product have been run (i.e. no new prefetch requests are coming)? Ok, the current configuration-driven mechanism doesn't guarantee that, but at least in principle.

Reply:
The last commit makes the resets only if `deleteEarly == false`, and also adds a unit test that fails if they would not be reset.